lib/events/events.go

```diff
@@ -351,3 +353,54 @@ func Error(err error) *string {
 	str := err.Error()
 	return &str
 }
+
+// Wait for the given event to become available, or for a timeout, whichever is sooner.
+// Returns true if the event became available, or false if we timed out.
+func (s *bufferedSubscription) waitForNewEventWithTimeout(id int, timeout time.Duration) bool {
+	// Checks whether 'id' is available. If it is, returns nil. If it isn't, returns newEventNotification (constructing it if necessary)
+	checkIDAndConstructNotificationIfNecessary := func() <-chan struct{} {
```

Comment:
I could, for example imagine a fairly clean TimeoutCond that does something like (pseudo code)

```
wait(timeout):
  - create a channel
  - with locking, add the channel to a list of waiting channels
  - select on the channel and a timeout

broadcast:
  - iterate over the list of waiting channels
    - close each
```